src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
+        private readonly LinkedListQueue<byte[]> _queue;
+
+        /// <summary>
+        /// Gets the input stream.
+        /// </summary>
+        /// <value>The input stream.</value>
+        public Stream InputStream { get; private set; }
+
+        /// <summary>
+        /// Gets the output stream.
+        /// </summary>
+        /// <value>The output stream.</value>
+        public Stream OutputStream { get; private set; }
+
+        public Pipe()
+        {
+            _queue = new LinkedListQueue<byte[]>();
+            InputStream = new PipeInputStream(_queue);
```

Comment:
If you like the non-NIO Java model better, I can refactor this into that and make input stream creation on-demand.